pyirf/io/gadf.py

```diff
@@ -303,3 +309,201 @@ def create_rad_max_hdu(
+        second axis is the offset angle, rest of the axes are IRF-specific
+    theta_bins: astropy.units.Quantity[angle]
+        theta bins
+    """
+
+    # to allow the function work on either single file or a list of files convert a single string into a list
+    if isinstance(files, str):
+        files = [files]
+
+    n_files = len(files)
+
+    irfs_all = list()
+
+    for this_file in files:
+        # [0] because there the IRFs are written as a single row of the table
+        irfs_all.append(QTable.read(this_file, hdu=extname)[field_name][0])
+
+    # if the function is run on single file do not need the first axis dimension
+    if n_files == 1:
+        irfs_all = irfs_all[0]
+    # the last operation converts the list to a multidimentional table
+    return np.array(irfs_all)
+
+
+def read_aeff2d_hdu(file_name, extname="EFFECTIVE AREA"):
+    """
+    Reads effective area from FITS file
+
+    Parameters
+    ----------
+    file_name: string or list of strings
+        file(s) to be read
+    extname:
+        Name for BinTableHDU
+
+    Returns
+    -------
+    effective_area: astropy.units.Quantity[area]
+        Effective area array of shape (n_energy_bins, n_fov_offset_bins) or (n_files, n_energy_bins, n_fov_offset_bins)
+    true_energy_bins: astropy.units.Quantity[energy]
+        Bin edges in true energy
+    fov_offset_bins: astropy.units.Quantity[angle]
+        Bin edges in the field of view offset.
+    """
+
+    field_name = "EFFAREA"
+    effective_area = read_irf_grid(file_name, extname, field_name)
+    true_energy_bins, fov_offset_bins = read_fits_bins_lo_hi(file_name, extname, ["ENERG", "THETA"])
+    true_energy_bins = binning.join_bin_lo_hi(*true_energy_bins)
+    fov_offset_bins = binning.join_bin_lo_hi(*fov_offset_bins)
+    return effective_area, true_energy_bins, fov_offset_bins
+
+
+def read_energy_dispersion_hdu(file_name, extname="EDISP"):
+    """
+    Reads energy dispersion table from a FITS file
+
+    Parameters
+    ----------
+    file_name: string or list of strings
+        file(s) to be read
+    extname:
+        Name for BinTableHDU
+
+    Returns
+    -------
+    energy_dispersion: numpy.ndarray
+        Energy dispersion array, with shape
+        (n_energy_bins, n_migra_bins, n_source_offset_bins)
+        or (n_files, n_energy_bins, n_migra_bins, n_source_offset_bins)
+    true_energy_bins: astropy.units.Quantity[energy]
+        Bin edges in true energy
+    migration_bins: numpy.ndarray
+        Bin edges for the relative energy migration (``reco_energy / true_energy``)
+    fov_offset_bins: astropy.units.Quantity[angle]
+        Bin edges in the field of view offset.
+    """
+
+    field_name = "MATRIX"
+    energy_dispersion = read_irf_grid(file_name, extname, field_name)
+    last_axis = len(energy_dispersion.shape) - 1
+    energy_dispersion = np.swapaxes(energy_dispersion, last_axis - 2, last_axis)
```

Comment:
Why swapping axis? That looks wrong too me. The only thing needed should be the transpose when reading from the table, not any swapaxes stuff.

Reply:
I think it is the same - transposition is exactly swapping axes. Please note that this is a 4D array, and you only need to swap two of them. I guess  .transpose(0, 1, 3, 2) would also work here, but swapaxis looks more clear to me here.

Reply:
No. The point is, transposing without any arguments should be done! Not just swapping two axis.

Could it be that the file your are testing this on does not follow the "Recommended axis order"?

Reply:
the test is done on prod 3 MCs, but as far as I remember it had the same axis sequence as the files saved  with pyirf.
Nevertheless, as we discussed in e-mail, I will update the tests to make them round tests so I will check it again.
Nevertheless I'm sure that it cannot be simple transpose, it would turn the axes 0, 1, 2, 3 into 3, 2, 1, 0 and this would not be correct.

Reply:
The recommended axis order is `ENERGY, MIGRA, THETA`, which is what we do here:

https://github.com/cta-observatory/pyirf/blob/0e7fcd29297a9ae95910f736bddcb73bb9cc1066/pyirf/irf/energy_dispersion.py#L67-L70

And we simply transpose to meet FITS conventions when exporting the HDU here:

https://github.com/cta-observatory/pyirf/blob/0e7fcd29297a9ae95910f736bddcb73bb9cc1066/pyirf/io/gadf.py#L185

Reply:
Thx @maxnoe indeed there was a problem, but not with migration matrix (which was read correctly), but with effective area, where messed up the axis order during refactoring of the code. Now I do the .T in read_irf_grid and it works fine for both Aeff and migration matrix